### DISCUSSION TITLE
Trying to distribute kotlin-cli-starter

### DISCUSSION THREAD

**Discussion by jmfayard** [hydraulic-software/conveyor#18] [General]
Hello I'm the  author of https://github.com/jmfayard/kotlin-cli-starter

Mike Hearn pinged me to try conveyor

I just finished the tutorial and so far it looks awesome

Do you have instructions on how to use the "Default GUI" so that I can distribute my CLI tool on windows/macOS?

Also how will updates work for a CLI tool? Won't they be only applied when open the "Default GUI" which people will never do?

**Comment by mikehearn** [comment]
Hi Jean-Michel,

That's a really cool project! Conveyor is written in Kotlin and uses Mordant, so this sort of thing is close to our hearts :) 

There are a few issues we'll have to work through before Conveyor is ready for your template, which is why the CLI support isn't fully documented yet.

## Kotlin Multiplatform

Two issues here:

1. KMP isn't properly supported by the Conveyor Gradle plugin, because JetBrains view KMP as alpha. Or did up until very recently, I think they now consider it as in beta. The KMP Kotlin plugin is quite different to the usual JVM plugin, different syntax, API etc.
2. We haven't yet tested Kotlin/Native apps.  It will probably require an extra flag to be used on macOS, something like `-linker-option﻿
 -header_pad 0xFF`.

The Conveyor Gradle plugin is open source - you can find it in this repository - so if you or anyone else wants to tackle this then that'd be cool. Upgrading it will require a bit of design work because in a KMP project what to package is ambiguous. Do you want to package the native build? The JVM build? Maybe even in future the JS build with Node? At the moment the Gradle support assumes there's one thing to package per project where the plugin is applied. We probably need to detect KMP projects and add new tasks like `printConveyorConfigForNative`, `printConveyorConfigForJVM` and so on.

For experimentation the Gradle plugin isn't needed. It just emits config, so you could come up with another way to generate that.

## Default GUI

To play with the default GUI use a config like this:

```
include required("/stdlib/jvm/default-gui.conf")

app {
  # ... usual stuff ...

  jvm {
    cli.your-program.main-class = com.example.YourProgram
  }
}
```

It's undocumented for a reason though - this feature isn't quite ready for primetime. In particular, it's currently for JVM apps only. To make the feature public we'll need to make sure it supports native apps too (Electron doesn't have any use for it). That probably means compiling the JAR with native-image but we haven't tried that yet.

## When to update?

The `default-gui.conf` will disable automatic background updates as well as enabling the default GUI. So indeed, users have to run the default GUI and click "Check for updates" to get upgraded. Is it wrong? Well, it encodes an assumption that users don't want CLI apps upgrading unless explicitly requested. That's the model used by homebrew, apt, yum/dnf, etc. It seemed like the path of least surprise.

This may not be correct. One of the things that we should do is split this config up, so the automatic update policy is separate from whether you get the default GUI. 

At the moment you're expected to alert the user that there's an upgrade available yourself. The next Conveyor release generates a `metadata.properties` file as part of the site output which contains the version, and in fact you can export near arbitrary config to this file, which is just a set of `key=value` lines so it's easy to parse from C++, Kotlin/Native etc. So by polling this file you can find out if there's an upgrade available.

The right way to do this depends on the app. For example it would be wrong to print such an alert if the output is redirected to a file, because such a redirection implies a script is processing the output and it doesn't want this sort of interference. It might also be wrong to poll for updates in the background, as maybe the user doesn't expect a CLI tool to contact the internet at random intervals like that, and maybe your CLI tool doesn't run long enough for that to make sense. In that case something like a date based reminder or an explicit command to upgrade would be appropriate. Or perhaps you want to just skip all that and just let the system package manager upgrade you.

Conveyor currently has a custom "message of the day" feature (that isn't being used right now), which lets us print a message when it's out of date. There's also an undocumented Java API to check for updates and trigger an update UI, which is what the default GUI uses. It's changed in v2 and will probably change again before being launched publicly, but what we can do earlier is expose a command line switch on the default GUI to run an update check without displaying anything unless an update is actually available. Update checks do have to run via a GUI unfortunately because Sparkle requires a Cocoa event loop to operate. On Linux of course it's all delegated to the system package manager, and on Windows the admin can use the package manager PowerShell cmdlets to do upgrades. It'd make sense to integrate with winget there too, and with homebrew on macOS. But it's not the highest priority right now.
